src/libraries/System.Net.Security/src/System/Net/Security/NetSecurityTelemetry.cs

```diff
@@ -0,0 +1,288 @@
+
+            long dummy = 0;
+            ref long connectionsOpen = ref dummy;
+            EventCounter? handshakeDurationCounter = null;
+
+            switch (protocol)
+            {
+                case SslProtocols.Tls:
+                    connectionsOpen = ref _connectionsOpenTls10;
+                    handshakeDurationCounter = _handshakeDurationTls10Counter;
+                    break;
+
+                case SslProtocols.Tls11:
+                    connectionsOpen = ref _connectionsOpenTls11;
+                    handshakeDurationCounter = _handshakeDurationTls11Counter;
+                    break;
+
+                case SslProtocols.Tls12:
+                    connectionsOpen = ref _connectionsOpenTls12;
+                    handshakeDurationCounter = _handshakeDurationTls12Counter;
+                    break;
+
+                case SslProtocols.Tls13:
+                    connectionsOpen = ref _connectionsOpenTls13;
+                    handshakeDurationCounter = _handshakeDurationTls13Counter;
+                    break;
+            }
+
+            if (connectionOpen)
+            {
+                Interlocked.Increment(ref connectionsOpen);
+            }
+
+            if (handshakeDurationCounter != null)
+            {
+                handshakeDurationCounter.WriteMetric(stopwatch.GetElapsedTime().TotalMilliseconds);
+            }
+
+            if (IsEnabled(EventLevel.Informational, EventKeywords.None))
+            {
+                HandshakeStop(protocol.ToString());
+            }
+        }
+
+        [NonEvent]
+        public void ConnectionClosed(SslProtocols protocol)
+        {
+            long count = 0;
+
+            switch (protocol)
+            {
+                case SslProtocols.Tls:
+                    count = Interlocked.Decrement(ref _connectionsOpenTls10);
+                    break;
+
+                case SslProtocols.Tls11:
+                    count = Interlocked.Decrement(ref _connectionsOpenTls11);
+                    break;
+
+                case SslProtocols.Tls12:
+                    count = Interlocked.Decrement(ref _connectionsOpenTls12);
+                    break;
+
+                case SslProtocols.Tls13:
+                    count = Interlocked.Decrement(ref _connectionsOpenTls13);
+                    break;
+            }
+
+            Debug.Assert(count >= 0);
+        }
+
+
+        [NonEvent]
+        private unsafe void WriteEvent(int eventId, bool arg1, string? arg2)
+        {
+            if (IsEnabled())
+            {
+                arg2 ??= string.Empty;
+
+                fixed (char* arg2Ptr = arg2)
+                {
+                    const int NumEventDatas = 2;
+                    EventData* descrs = stackalloc EventData[NumEventDatas];
+
+                    descrs[0] = new EventData
+                    {
+                        DataPointer = (IntPtr)(&arg1),
+                        Size = sizeof(int) // EventSource defines bool as a 32-bit type
```

Comment:
Is this the right way of using bools in event data?

Reply:
See https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs#L867 for the right way to encode `bool` types.  Also, I believe that initlocals is not set for this assembly, so you will need to explicitly set `EventData.Reserved=0`.  A non-zero value will cause failures that will result in the events being corrupted.

Given that the input types are all supported types, you could delete this entire method and just let `EventSource` handle it for you.  It's OK to do this, but there's the potential for more bugs here.

Also, given that these are

Reply:
> Also, I believe that initlocals is not set for this assembly, so you will need to explicitly set EventData.Reserved=0. A non-zero value will cause failures that will result in the events being corrupted.

`Reserved` is not visible outside of its assembly. Using `new EventData` will zero-init that field, regardless of `SkipLocalsInit`.

> Given that the input types are all supported types, you could delete this entire method and just let EventSource handle it for you.

Considering these events are only fired 2/3x per connection, the perf overhead likely wouldn't matter.

Reply:
That's a good point - you're using the constructor here, so you should be fine.